diff --git a/m4/Makefile.am.targets b/m4/Makefile.am.targets
index 9e39175..20cbf81 100644
--- a/m4/Makefile.am.targets
+++ b/m4/Makefile.am.targets
@@ -67,7 +67,7 @@ $(CMXPK): $(CMX)
 # META description for package
 META: $(top_srcdir)/VERSION
 	@echo "Generating META description"; \
-	revision=`grep "\\$$Revision:" $< | sed -e 's,^\\$$,,' -e 's, \\$$$$,,'`; \
+	revision=`cat $<`; \
 	date=`LC_TIME=C date`; \
 	echo "version = \"$$revision\"" > $@; \
 	echo "description = \"$(DESCRIPTION)\"" >> $@; \
